File: liqp-core/src/main/java/liqp/nodes/LookupNode.java
package liqp.nodes;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import liqp.node.LNode;
import liqp.exceptions.MissingVariableException;
import liqp.lookup.Indexable;
import lombok.Getter;

@Getter
public class LookupNode implements LNode {

  private final String id;
  private final List<Indexable> indexes;

  public LookupNode(String id) {
    this.id = id;
    indexes = new ArrayList<>();
  }

  public void add(Indexable indexable) {
    indexes.add(indexable);
  }

  @Override
  public Object render(LContext context) {

    Object value;

    // Check if there's a [var] lookup, AST: ^(LOOKUP Id["@var"])
    if (id.startsWith("@")) {
      final Object existing = context.get(id.substring(1));
      final String varName = String.valueOf(existing);
      value = context.get(varName);
    } else {
      value = context.get(id);
    }

    try {
      for (Indexable index : indexes) {
        value = index.get(value, context);
      }
    } catch (MissingVariableException e) {
      throw new MissingVariableException(this.getVariableName(), e.getVariableName());
    }

    return value;
  }

  @Override
  public String toString() {

    StringBuilder builder = new StringBuilder();

    builder.append(id);

    for (Indexable index : this.indexes) {
      builder.append(index.toString());
    }

    return builder.toString();
  }

  private String getVariableName() {
    StringBuilder variableFullName = new StringBuilder(id);
    for (Indexable index : indexes) {
      variableFullName.append(index.toString());
    }
    return variableFullName.toString();
  }

  @Override
  public List<LNode> children() {
    return Collections.emptyList();
  }
}

File: liqp-core/src/main/kotlin/liqp/nodes/LookupNode.kt
package liqp.nodes

import liqp.context.LContext
import liqp.exceptions.MissingVariableException
import liqp.lookup.Indexable
import liqp.node.LNode
import liqp.resolve

class LookupNode(val id: String,
                 private val indexes: List<Indexable> = emptyList()) : LNode() {

  private val variableName: String by lazy {
    val variableFullName = StringBuilder(id)
    for (index in indexes) {
      variableFullName.append(index.toString())
    }
    variableFullName.toString()
  }

  override val children: List<LNode> = emptyList()
  override fun render(context: LContext): Any? {
    // Check if there's a [var] lookup, AST: ^(LOOKUP Id["@var"])
    var value: Any? = when {
      id == "_" -> context.inputData
      id.startsWith("@") -> {
        val existing: Any? = context[id.substring(1)]
        val varName = existing.toString()
        context[varName]
      }
      else -> context[id]
    }

    if (value == null && context.parseSettings.isStrictVariables) {
      throw MissingVariableException(variableName)
    }

    try {
      value = indexes.resolve(value, context)
    } catch (e: MissingVariableException) {
      throw MissingVariableException(this.variableName, e.variableName)
    }

    return value
  }

  override fun toString(): String {
    return variableName
  }
}
